src/main/scala/io/iohk/ethereum/jsonrpc/EthService.scala

```diff
@@ -146,17 +156,44 @@ class EthService(blockchain: Blockchain, blockGenerator: BlockGenerator, appStat
 
   def getWork(req: GetWorkRequest): ServiceResponse[GetWorkResponse] = {
     import io.iohk.ethereum.mining.pow.PowCache._
-    val block = blockGenerator.generateBlockForMining()
-    Future.successful(Right(GetWorkResponse(
-      powHeaderHash = ByteString(kec256(BlockHeader.getEncodedWithoutNonce(block.header))),
-      dagSeed = seedForBlock(block.header.number),
-      target = ByteString((BigInt(2).pow(256) / block.header.difficulty).toByteArray)
-    )))
+
+    val blockNumber = appStateStorage.getBestBlockNumber() + 1
+
+    import scala.concurrent.ExecutionContext.Implicits.global
+    implicit val timeout = Timeout(5 seconds)
+    (pendingTransactionsManager ? PendingTransactionsManager.GetPendingTransactions).map{
+      case PendingTransactions(txList) =>
+        val fakeAddress = 42
+        //todo ask for ommers
+        val ommersList = Nil
+        val block = blockGenerator.generateBlockForMining(blockNumber, txList, ommersList, Address(fakeAddress))
+
+        block match {
+          case Right(b) =>
+            Right(GetWorkResponse(
+              powHeaderHash = ByteString(kec256(BlockHeader.getEncodedWithoutNonce(b.header))),
+              dagSeed = seedForBlock(b.header.number),
+              target = ByteString((BigInt(2).pow(256) / b.header.difficulty).toByteArray)
+            ))
+          case Left(err) =>
+            log.error(s"unable to prepare block because of $err")
+            Left(JsonRpcErrors.InternalError)
+        }
+      case a =>
+        log.error(s"unable to get pending transactions")
+        Left(JsonRpcErrors.InternalError)
+    }
   }
 
   def submitWork(req: SubmitWorkRequest): ServiceResponse[SubmitWorkResponse] = {
-    //todo add logic for including mined block into blockchain
-    Future.successful(Right(SubmitWorkResponse(true)))
+    blockGenerator.getPrepared(req.powHeaderHash) match {
+      case Some(block) if appStateStorage.getBestBlockNumber() <= block.header.number =>
+        syncingController ! MinedBlock(block.copy(header = block.header.copy(nonce = req.nonce, mixHash = req.mixHash)))
+        pendingTransactionsManager ! PendingTransactionsManager.RemoveTransactions(block.body.transactionList)
```

Comment:
yes they are, will check if we can prevent that, but it is not trivial because then we would add transactions when doing reorganization of the chain which may led to inconsistent transaction pool
for example:
while reorganizing chain we removed some transactions because they were included in blocks that we got from peer and then we include them because we received mined block while reorg in that case we should somehow filter transactions from ours newly mined block